### PR TITLE
Enh save lists quick

### PR DESCRIPTION
**Pull request by to266**
The numpy solution, suggested in #642.

Currently will fail for 0d data, as it has not been fixed in the upstream, however two possible fixes are:
#644, #645

### CODE REVIEW

**Comment by francisco-dlp** [comment]
On loading lists become arrays:

```python
>>> import hyperspy.hspy as hs
>>> s = hs.signals.Signal([0])
>>> s.metadata.test = [1, 2, 3]
>>> s.save("test")
The test.hdf5 file was created
>>> load("test.hdf5").metadata.test
array([1, 2, 3])

```

Is this intended?

File: hyperspy/io_plugins/hdf5.py [hunk 1]
```diff
@@ -339,20 +339,40 @@ def dict2hdfgroup(dictionary, group, compression=None):
             group.attrs["_datetime_" + key] = repr(value)
         elif isinstance(value, list):
```
Comment (francisco-dlp) [inline hyperspy/io_plugins/hdf5.py]: It seems that a good fraction of the code for saving lists and tuples is the same, wouldn't it be possible to combine the two?

**Comment by to266** [comment]
Yes, as discussed in https://github.com/hyperspy/hyperspy/issues/642.

It's essentially the price we pay if we want to save them quickly

**Comment by francisco-dlp** [comment]
But, wouldn't it be possible to convert them back to list on loading?

File: hyperspy/io_plugins/hdf5.py (hunk 1 above)
Comment (to266) [inline hyperspy/io_plugins/hdf5.py]: It is possible, I was just lazy, as we have not decided we are going this way. So.. a PR with the example behaviour, but not quite ready for merge?

File: hyperspy/io_plugins/hdf5.py [hunk 2]
```diff
@@ -339,20 +339,40 @@ def dict2hdfgroup(dictionary, group, compression=None):
...
-                dict2hdfgroup(dict(zip(
-                    [unicode(i) for i in xrange(len(value))], value)),
-                    group.create_group(
-                    '_list_' + str(len(value)) + '_' + key),
-                    compression=compression)
+                try:
+                    tmp = np.array(value)
+                except ValueError:
+                    tmp = np.array([[0]])
+                if tmp.dtype is np.dtype('O') or tmp.ndim is not 1:
```
Comment (francisco-dlp) [inline hyperspy/io_plugins/hdf5.py]: What's wrong with ndim > 1?

**Comment by to266** [comment]
It is possible, sure.
However I think there is no point in converting them back - most of the time you'll want to perform array operations any way. 

Also, if we leave the structures as numpy arrays, it may be somewhat clearer for the user why all floats became strings and similar. If instead we converted them back to regular lists and tuples, then it gets hidden.

File: hyperspy/io_plugins/hdf5.py (hunk 2 above)
Comment (to266) [inline hyperspy/io_plugins/hdf5.py]: For consistency. If no such exception is made, then [(1,2),(3,4)] will get turned into one array of ndim=2, but [(1,2), (3,4,5)] - into a list of two arrays, each of ndim=1. So here I make sure it's always one-dimensional. Effectively preventing numpy from "eating up" not only the attribute types (float, string), but also outer layer type (tuple, list)

**Comment by francisco-dlp** [comment]
I see your point. However, I think that most people would expect to get back exactly what they've saved. Otherwise, one can write a script that works on some data, save the data in hdf5, load it and find that the script is broken when operating on the newly loaded data.

File: hyperspy/io_plugins/hdf5.py (hunk 2 above)
Comment (francisco-dlp) [inline hyperspy/io_plugins/hdf5.py]: Very good point. Could you add a little comment to the code to explain this?

**Comment by to266** [comment]
But it will not be the same any way - as I've showed in https://github.com/hyperspy/hyperspy/issues/642, floats will be turned to strings, etc

**Comment by francisco-dlp** [comment]
Ok, but fixing that other (possibly rare) issue is not straightforward, while fixing the (far more common) list/tuple becomes array issue should be beyond reach.

By the way, we should not forget to warn the users in the docs about numpy's homogenizing efforts and the consequences for their data.

**Comment by to266** [comment]
Ok, a fair point. So do we look for other solutions for #642, or go this way? If we do, I'll add the missing bits and clean up the code somewhat, as it's not ready for merge now.

**Comment by francisco-dlp** [comment]
I'm +1 on this solution.

**Pull request merged**